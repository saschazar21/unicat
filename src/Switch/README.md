Renders a boolean Switch:

```jsx
<Switch name="hello" />
```

To render a pre-checked version, append a `checked` prop:

```jsx
<Switch name="hello" checked />
```

Also, a `large` variant is possible:

```jsx
<Switch name="hello" large />
```
