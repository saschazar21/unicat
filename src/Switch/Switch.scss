@import '../__styles__/variables';

.wrapper {
  display: inline-block;
  position: relative;
  height: 1em;
  width: 2em;
}

.large {
  height: 1.5em;
  width: 3em;
}

.srOnly {
  position: absolute;
  border: 0;
  height: 1px;
  width: 1px;
  padding: 0;
  margin: -1px;
  clip: rect(0, 0, 0, 0);
  overflow: hidden;
}

.circle {
  transition: transition(left);
  position: absolute;
  border-radius: borderRadius(2);
  box-shadow: boxShadow(m);
  top: 50%;
  left: borderWidth(2);
  height: calc(100% - #{borderWidth(2) * 2});
  width: calc(50% - #{borderWidth(2) * 2});
  transform: translateY(-50%);

  .input:checked + .container > & {
    left: calc(50% + #{borderWidth(2)});
  }
}

.container {
  transition: transition(background-color);
  position: absolute;
  border-radius: borderRadius(2);
  box-shadow: boxShadow(xs, true);
  background-color: color(grey, false, $systemColors);
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  overflow: hidden;

  > .circle {
    fill: color(default, true, $systemColors);
  }

  .input:checked + & {
    background-color: color(light);

    > .circle {
      fill: color(white, false, $systemColors);
    }
  }

  .input:focus + & {
    @include focus();
  }
}

.input {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  width: 100%;
  opacity: 0;
  z-index: 10;

  &:active,
  &:hover {
    cursor: pointer;
  }
}
