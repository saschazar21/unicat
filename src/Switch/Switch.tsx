import React, { useState, KeyboardEvent, SyntheticEvent } from 'react';
import classnames from 'classnames';

import { preventDefault } from '../__tools__/helpers';

import styles from './Switch.scss';

export interface SwitchProps {
  /* rendered as checked */
  checked?: boolean;
  /* custom class name */
  className?: string;
  /* set an ID for a label */
  id?: string;
  /* grow Switch larger */
  large?: boolean;
  /* set a name */
  name: string;
  /* onClick handler */
  onClick?: (event: SyntheticEvent) => void;
  /* custom prefix */
  prefix?: string;
  /* custom suffix */
  suffix?: string;
}

export default function Switch(props: SwitchProps): JSX.Element {
  const {
    checked: initChecked,
    className: customClassName,
    large,
    name,
    onClick,
    prefix,
    suffix,
  } = props;
  const [checked, setChecked] = useState(!!initChecked);

  const className = classnames(
    { [styles.verbose]: prefix || suffix },
    styles.wrapper,
    customClassName
  );

  const handleClick = (event: SyntheticEvent): void => {
    setChecked(prevState => !prevState);

    if (onClick) {
      onClick(event);
    }
  };

  const handleKeyUp = (event: KeyboardEvent<HTMLInputElement>): void => {
    preventDefault(event);
    if (event.key === ' ' || event.key === 'Enter') {
      handleClick(event);
    }
  };

  return (
    <div
      className={className}
      data-checked={checked}
      data-prefix={prefix}
      data-suffix={suffix}
    >
      <input
        className={styles.input}
        type="checkbox"
        name={name}
        checked={checked}
        onClick={handleClick}
        onKeyUp={handleKeyUp}
      />
      <div
        className={classnames({ [styles.large]: large }, styles.container)}
        aria-hidden="true"
      >
        <svg viewBox="0 0 16 16" className={styles.circle}>
          <circle cx="8" cy="8" r="8" />
          <circle cx="8" cy="8" r="6" />
        </svg>
      </div>
    </div>
  );
}

Switch.displayName = 'Switch';
