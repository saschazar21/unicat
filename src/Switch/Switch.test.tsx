import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Switch from './Switch';

describe('Switch', () => {
  it('renders', () => {
    const wrapper = shallow(<Switch name="test" />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders a large variant', () => {
    const wrapper = shallow(<Switch name="test" large />);

    expect(toJson(wrapper)).toMatchSnapshot();

    expect(wrapper.find('[checked="false"]')).toHaveLength(0);

    wrapper.find('input').simulate('click');

    expect(wrapper.find('[checked=true]')).toHaveLength(1);
  });

  it('renders a checked variant', () => {
    const wrapper = shallow(<Switch name="test" checked />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders a verbose variant', () => {
    const wrapper = shallow(<Switch name="test" verbose />);

    expect(wrapper.find('.verbose')).toHaveLength(1);
  });

  it('handles click', () => {
    const onClick = jest.fn();

    const wrapper = shallow(<Switch name="test" onClick={onClick} />);

    expect(onClick).not.toHaveBeenCalled();

    wrapper.find('input').simulate('click');

    expect(onClick).toHaveBeenCalled();
  });
});
