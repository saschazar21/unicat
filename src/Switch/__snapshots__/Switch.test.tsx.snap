// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders 1`] = `
<div
  className="wrapper"
  data-checked={false}
>
  <input
    checked={false}
    className="input"
    name="test"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <div
    aria-hidden="true"
    className="container"
  >
    <svg
      className="circle"
      viewBox="0 0 16 16"
    >
      <circle
        cx="8"
        cy="8"
        r="8"
      />
      <circle
        cx="8"
        cy="8"
        r="6"
      />
    </svg>
  </div>
</div>
`;

exports[`Switch renders a checked variant 1`] = `
<div
  className="wrapper"
  data-checked={true}
>
  <input
    checked={true}
    className="input"
    name="test"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <div
    aria-hidden="true"
    className="container"
  >
    <svg
      className="circle"
      viewBox="0 0 16 16"
    >
      <circle
        cx="8"
        cy="8"
        r="8"
      />
      <circle
        cx="8"
        cy="8"
        r="6"
      />
    </svg>
  </div>
</div>
`;

exports[`Switch renders a large variant 1`] = `
<div
  className="wrapper"
  data-checked={false}
>
  <input
    checked={false}
    className="input"
    name="test"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <div
    aria-hidden="true"
    className="large container"
  >
    <svg
      className="circle"
      viewBox="0 0 16 16"
    >
      <circle
        cx="8"
        cy="8"
        r="8"
      />
      <circle
        cx="8"
        cy="8"
        r="6"
      />
    </svg>
  </div>
</div>
`;
